libgadget/slotsmanager.c

```diff
@@ -424,21 +402,84 @@ slots_get_last_garbage(int nfirst, int nlast, int ptype, const struct part_manag
     return nlast;
 }
 
+struct PeanoOrder
+{
+    peano_t Key;
+    int Type;
+    int Pindex;
+};
+
+static int
+order_by_type_and_key(const void *a, const void *b)
+{
+    const struct PeanoOrder * pa  = (const struct PeanoOrder *) a;
+    const struct PeanoOrder * pb  = (const struct PeanoOrder *) b;
+    /* Note garbage types have their values set to something large here*/
+    if(pa->Type < pb->Type)
+        return -1;
+    if(pa->Type > pb->Type)
+        return +1;
+    if(pa->Key < pb->Key)
+        return -1;
+    if(pa->Key > pb->Key)
+        return +1;
+
+    return 0;
+}
+
 /* Sort the particles and their slots by type and peano order.
  * This does a gc by sorting the Garbage to the end of the array and then trimming.
  * It is a different algorithm to slots_gc, somewhat slower,
  * but delivers a spatially compact sort. It always compacts the slots*/
 void
 slots_gc_sorted(struct part_manager_type * pman, struct slots_manager_type * sman)
 {
-    int ptype;
+    int ptype, i;
     /* Resort the particles such that those of the same type and key are close by.
      * The locality is broken by the exchange. */
-    qsort_openmp(pman->Base, pman->NumPart, sizeof(struct particle_data), order_by_type_and_key);
-
+    int64_t garbage=0;
+    struct PeanoOrder * peanokeys = mymalloc("Keydata", pman->NumPart * sizeof(struct PeanoOrder));
+    #pragma omp parallel for reduction(+: garbage)
+    for(i = 0; i < pman->NumPart; i++) {
+        peanokeys[i].Key = PEANO(pman->Base[i].Pos, pman->BoxSize);
+        peanokeys[i].Type = pman->Base[i].Type;
+        if(pman->Base[i].IsGarbage) {
+            garbage++;
+            /* Set it to a type far beyond the existing types, so it sorts to the end.*/
+            peanokeys[i].Type = 255;
```

Comment:
Since Type in peanokeys differs from Type (a value of 255 for garbage), perhaps rename the field to TypeKey?

Reply:
Done!